// lucid-ui/dist/components/SearchField/SearchField.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});
exports.SearchFieldDumb = undefined;

var _get2 = require('lodash/get');

var _get3 = _interopRequireDefault(_get2);

var _isEmpty2 = require('lodash/isEmpty');

var _isEmpty3 = _interopRequireDefault(_isEmpty2);

var _isUndefined2 = require('lodash/isUndefined');

var _isUndefined3 = _interopRequireDefault(_isUndefined2);

var _noop2 = require('lodash/noop');

var _noop3 = _interopRequireDefault(_noop2);

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('react-peek/prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _styleHelpers = require('../../util/style-helpers');

var _componentTypes = require('../../util/component-types');

var _stateManagement = require('../../util/state-management');

var _TextField = require('../TextField/TextField');

var _TextField2 = _interopRequireDefault(_TextField);

var _SearchIcon = require('../Icon/SearchIcon/SearchIcon');

var _SearchIcon2 = _interopRequireDefault(_SearchIcon);

var _SearchField = require('./SearchField.reducers');

var _SearchField2 = _interopRequireDefault(_SearchField);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var cx = _styleHelpers.lucidClassNames.bind('&-SearchField');

var bool = _propTypes2.default.bool,
    func = _propTypes2.default.func,
    node = _propTypes2.default.node,
    number = _propTypes2.default.number,
    oneOfType = _propTypes2.default.oneOfType,
    string = _propTypes2.default.string;


var SearchField = (0, _componentTypes.createClass)({
	displayName: 'SearchField',

	statics: {},

	components: {
		TextField: _TextField2.default,
		Icon: (0, _componentTypes.createClass)({
			displayName: 'SearchField.Icon',
			statics: {},
			propName: 'Icon'
		})
	},

	reducers: _SearchField2.default,

	propTypes: {
		onChange: func,

		onChangeDebounced: func,

		debounceLevel: number,

		onSubmit: func,

		value: oneOfType([number, string]),

		isValid: bool,

		isDisabled: bool,

		placeholder: string,

		className: string,

		Icon: node
	},

	getDefaultProps: function getDefaultProps() {
		return {
			isDisabled: false,
			onChange: _noop3.default,
			onChangeDebounced: _noop3.default,
			debounceLevel: 500,
			onSubmit: _noop3.default,
			value: ''
		};
	},
	render: function render() {
		var props = this.props,
		    _props = this.props,
		    className = _props.className,
		    isDisabled = _props.isDisabled,
		    isValid = _props.isValid,
		    onChange = _props.onChange,
		    onChangeDebounced = _props.onChangeDebounced,
		    debounceLevel = _props.debounceLevel,
		    onSubmit = _props.onSubmit,
		    placeholder = _props.placeholder,
		    value = _props.value,
		    passThroughs = _objectWithoutProperties(_props, ['className', 'isDisabled', 'isValid', 'onChange', 'onChangeDebounced', 'debounceLevel', 'onSubmit', 'placeholder', 'value']);

		var Icon = SearchField.Icon,
		    TextField = SearchField.TextField;


		var textFieldProps = {
			isDisabled: isDisabled,
			onChange: onChange,
			onChangeDebounced: onChangeDebounced,
			debounceLevel: debounceLevel,
			onSubmit: onSubmit,
			placeholder: placeholder,
			isMultiLine: false,
			value: value
		};

		var textFieldElement = (0, _componentTypes.getFirst)(props, TextField) || _react2.default.createElement(TextField, textFieldProps);
		var isIconActive = (0, _isUndefined3.default)(isValid) ? !(0, _isEmpty3.default)((0, _get3.default)(textFieldElement, 'props.value')) : isValid;
		var defaultIcon = _react2.default.createElement(_SearchIcon2.default, { className: cx('&-Icon', { '&-Icon-active': isIconActive }) });
		var iconElement = (0, _componentTypes.getFirst)(props, Icon);
		var iconChildren = (0, _get3.default)(iconElement, 'props.children');
		var icon = iconChildren ? (0, _react.createElement)(iconChildren.type, _extends({}, iconChildren.props, {
			className: cx('&-Icon', { '&-Icon-active': isIconActive }, iconChildren.props.className)
		})) : defaultIcon;

		return _react2.default.createElement(
			'div',
			_extends({}, (0, _componentTypes.omitProps)(passThroughs, SearchField), {
				className: cx('&', className)
			}),
			textFieldElement,
			_react2.default.createElement(
				'div',
				{ className: cx('&-Icon-container') },
				icon
			)
		);
	}
});

exports.default = (0, _stateManagement.buildHybridComponent)(SearchField);
exports.SearchFieldDumb = SearchField;
