// lucid-ui/dist/components/Validation/Validation.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _get2 = require('lodash/get');

var _get3 = _interopRequireDefault(_get2);

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('react-peek/prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _styleHelpers = require('../../util/style-helpers');

var _componentTypes = require('../../util/component-types');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var cx = _styleHelpers.lucidClassNames.bind('&-Validation');

var any = _propTypes2.default.any;


var Validation = (0, _componentTypes.createClass)({
	displayName: 'Validation',

	statics: {},

	components: {
		Error: (0, _componentTypes.createClass)({
			displayName: 'Validation.Error',
			statics: {},
			propName: 'Error'
		})
	},

	propTypes: {
		Error: any,

		className: any,

		children: any.isRequired
	},

	render: function render() {
		var _props = this.props,
		    className = _props.className,
		    children = _props.children,
		    passThroughs = _objectWithoutProperties(_props, ['className', 'children']);

		var errorChildProps = (0, _get3.default)((0, _componentTypes.getFirst)(this.props, Validation.Error), 'props');

		return _react2.default.createElement(
			'div',
			_extends({}, (0, _componentTypes.omitProps)(passThroughs, Validation), {
				className: cx('&', {
					'&-is-error': errorChildProps && errorChildProps.children
				}, className)
			}),
			children,
			errorChildProps && errorChildProps.children && errorChildProps.children !== true ? _react2.default.createElement(
				'div',
				_extends({}, (0, _componentTypes.omitProps)(errorChildProps, Validation.Error), {
					className: cx('&-error-content', errorChildProps.className)
				}),
				errorChildProps.children
			) : null
		);
	}
});

exports.default = Validation;
