// lucid-ui/dist/components/TextFieldValidated/TextFieldValidated.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _map2 = require('lodash/map');

var _map3 = _interopRequireDefault(_map2);

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('react-peek/prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _Validation = require('../Validation/Validation');

var _Validation2 = _interopRequireDefault(_Validation);

var _TextField = require('../TextField/TextField');

var _TextField2 = _interopRequireDefault(_TextField);

var _TextField3 = require('../TextField/TextField.reducers');

var _TextField4 = _interopRequireDefault(_TextField3);

var _styleHelpers = require('../../util/style-helpers');

var _componentTypes = require('../../util/component-types');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var cx = _styleHelpers.lucidClassNames.bind('&-TextFieldValidated');

var any = _propTypes2.default.any,
    object = _propTypes2.default.object,
    string = _propTypes2.default.string;


var TextFieldValidated = (0, _componentTypes.createClass)({
	displayName: 'TextFieldValidated',

	statics: {},

	reducers: _TextField4.default,

	components: {
		Error: (0, _componentTypes.createClass)({
			displayName: 'TextFieldValidated.Error',
			statics: {},
			propName: 'Error'
		})
	},

	propTypes: {
		style: object,

		className: string,

		Error: any
	},

	getDefaultProps: function getDefaultProps() {
		return _TextField2.default.getDefaultProps();
	},
	focus: function focus() {
		this.refs.TextField.focus();
	},
	render: function render() {
		var _this = this;

		var _props = this.props,
		    className = _props.className,
		    style = _props.style,
		    passThroughs = _objectWithoutProperties(_props, ['className', 'style']);

		var errorChildProps = (0, _map3.default)((0, _componentTypes.findTypes)(this.props, TextFieldValidated.Error), 'props');

		return _react2.default.createElement(
			_Validation2.default,
			{
				className: cx('&', className),
				style: style,
				Error: errorChildProps
			},
			_react2.default.createElement(_TextField2.default, _extends({}, (0, _componentTypes.omitProps)(passThroughs, TextFieldValidated, [], false), {
				ref: function ref(_ref) {
					return _this.refs = { TextField: _ref };
				}
			}))
		);
	}
});

exports.default = TextFieldValidated;
