// lucid-ui/dist/components/Dialog/Dialog.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _keys2 = require('lodash/keys');

var _keys3 = _interopRequireDefault(_keys2);

var _pick2 = require('lodash/pick');

var _pick3 = _interopRequireDefault(_pick2);

var _get2 = require('lodash/get');

var _get3 = _interopRequireDefault(_get2);

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('react-peek/prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _Overlay = require('../Overlay/Overlay');

var _Overlay2 = _interopRequireDefault(_Overlay);

var _styleHelpers = require('../../util/style-helpers');

var _componentTypes = require('../../util/component-types');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var cx = _styleHelpers.lucidClassNames.bind('&-Dialog');

var node = _propTypes2.default.node,
    oneOf = _propTypes2.default.oneOf,
    bool = _propTypes2.default.bool;


var SMALL = 'small';
var MEDIUM = 'medium';
var LARGE = 'large';

var Dialog = (0, _componentTypes.createClass)({
	displayName: 'Dialog',

	statics: {},

	components: {
		Header: (0, _componentTypes.createClass)({
			displayName: 'Dialog.Header',
			statics: {},
			propName: 'Header'
		}),
		Footer: (0, _componentTypes.createClass)({
			displayName: 'Dialog.Footer',
			statics: {},
			propName: 'Footer'
		})
	},

	propTypes: _extends({}, _Overlay2.default.propTypes, {

		size: oneOf(['small', 'medium', 'large']),

		Header: node,

		Footer: node,

		hasGutters: bool
	}),

	getDefaultProps: function getDefaultProps() {
		return {
			size: MEDIUM,
			hasGutters: true
		};
	},
	render: function render() {
		var _props = this.props,
		    className = _props.className,
		    size = _props.size,
		    hasGutters = _props.hasGutters,
		    isShown = _props.isShown,
		    passThroughs = _objectWithoutProperties(_props, ['className', 'size', 'hasGutters', 'isShown']);

		var headerChildProp = (0, _get3.default)((0, _componentTypes.getFirst)(this.props, Dialog.Header), 'props', {});
		var footerChildProp = (0, _get3.default)((0, _componentTypes.getFirst)(this.props, Dialog.Footer), 'props', null);

		return _react2.default.createElement(
			_Overlay2.default,
			_extends({}, (0, _componentTypes.omitProps)(passThroughs, Dialog, [], false), (0, _pick3.default)(passThroughs, (0, _keys3.default)(_Overlay2.default.propTypes)), {
				isShown: isShown,
				className: cx('&', className)
			}),
			_react2.default.createElement(
				'div',
				{
					className: cx('&-window', {
						'&-window-is-small': size === SMALL,
						'&-window-is-medium': size === MEDIUM,
						'&-window-is-large': size === LARGE
					})
				},
				_react2.default.createElement('header', _extends({}, headerChildProp, { className: cx('&-header') })),
				_react2.default.createElement(
					'section',
					{
						className: cx('&-body', hasGutters ? '' : '&-body-no-gutters')
					},
					this.props.children
				),
				footerChildProp && _react2.default.createElement('footer', _extends({}, footerChildProp, { className: cx('&-footer') }))
			)
		);
	}
});

exports.default = Dialog;
