// lucid-ui/dist/components/Panel/Panel.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _map2 = require('lodash/map');

var _map3 = _interopRequireDefault(_map2);

var _first2 = require('lodash/first');

var _first3 = _interopRequireDefault(_first2);

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('react-peek/prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _styleHelpers = require('../../util/style-helpers');

var _componentTypes = require('../../util/component-types');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var cx = _styleHelpers.lucidClassNames.bind('&-Panel');

var bool = _propTypes2.default.bool,
    node = _propTypes2.default.node,
    object = _propTypes2.default.object,
    string = _propTypes2.default.string;


var Panel = (0, _componentTypes.createClass)({
	displayName: 'Panel',

	statics: {},

	propName: 'Panel',

	components: {
		Header: (0, _componentTypes.createClass)({
			displayName: 'Panel.Header',
			statics: {},
			propName: 'Header'
		}),
		Footer: (0, _componentTypes.createClass)({
			displayName: 'Panel.Footer',
			statics: {},
			propName: 'Footer'
		})
	},

	propTypes: {
		className: string,
		Header: node,
		Footer: node,
		children: node,
		isGutterless: bool,
		hasMargin: bool,
		style: object,
		isScrollable: bool
	},

	getDefaultProps: function getDefaultProps() {
		return {
			isGutterless: false,
			hasMargin: true,
			isScrollable: true
		};
	},


	render: function render() {
		var _props = this.props,
		    children = _props.children,
		    className = _props.className,
		    isGutterless = _props.isGutterless,
		    hasMargin = _props.hasMargin,
		    style = _props.style,
		    isScrollable = _props.isScrollable,
		    passThroughs = _objectWithoutProperties(_props, ['children', 'className', 'isGutterless', 'hasMargin', 'style', 'isScrollable']);

		var headerChildProp = (0, _first3.default)((0, _map3.default)((0, _componentTypes.findTypes)(this.props, Panel.Header), 'props'));
		var footerChildProp = (0, _first3.default)((0, _map3.default)((0, _componentTypes.findTypes)(this.props, Panel.Footer), 'props'));

		return _react2.default.createElement(
			'div',
			_extends({}, (0, _componentTypes.omitProps)(passThroughs, Panel), {
				className: cx('&', className, {
					'&-is-not-gutterless': !isGutterless,
					'&-has-margin': hasMargin,
					'&-is-scrollable': isScrollable
				}),
				style: style
			}),
			headerChildProp ? _react2.default.createElement('header', _extends({}, headerChildProp, {
				className: cx('&-Header', headerChildProp.className)
			})) : null,
			_react2.default.createElement(
				'section',
				{ className: cx('&-content') },
				children
			),
			footerChildProp ? _react2.default.createElement('footer', _extends({}, footerChildProp, {
				className: cx('&-Footer', footerChildProp.className)
			})) : null
		);
	}
});

exports.default = Panel;
