// lucid-ui/dist/components/ProgressBar/ProgressBar.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _get2 = require('lodash/get');

var _get3 = _interopRequireDefault(_get2);

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('react-peek/prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _styleHelpers = require('../../util/style-helpers.js');

var _componentTypes = require('../../util/component-types.js');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var cx = _styleHelpers.lucidClassNames.bind('&-ProgressBar');

var number = _propTypes2.default.number,
    string = _propTypes2.default.string,
    oneOf = _propTypes2.default.oneOf,
    node = _propTypes2.default.node;


var ProgressBar = (0, _componentTypes.createClass)({
	displayName: 'ProgressBar',

	statics: {},

	propName: 'ProgressBar',

	components: {
		Title: (0, _componentTypes.createClass)({
			displayName: 'ProgressBar.Title',
			statics: {},
			propName: 'Title'
		})
	},

	propTypes: {
		className: string,

		kind: oneOf(['default', 'success', 'danger', 'info', 'warning']),

		percentComplete: number,

		children: node,

		Title: node
	},

	getDefaultProps: function getDefaultProps() {
		return {
			kind: 'default',
			percentComplete: 0
		};
	},
	render: function render() {
		var _props = this.props,
		    kind = _props.kind,
		    percentComplete = _props.percentComplete,
		    className = _props.className,
		    passThroughs = _objectWithoutProperties(_props, ['kind', 'percentComplete', 'className']);

		var titleChildProp = (0, _get3.default)((0, _componentTypes.getFirst)(this.props, ProgressBar.Title), 'props', {});

		return _react2.default.createElement(
			'div',
			_extends({}, (0, _componentTypes.omitProps)(passThroughs, ProgressBar), {
				className: cx('&', className, {
					'&-default': kind === 'default',
					'&-success': kind === 'success',
					'&-danger': kind === 'danger',
					'&-info': kind === 'info',
					'&-warning': kind === 'warning'
				})
			}),
			_react2.default.createElement('title', _extends({}, titleChildProp, { className: cx('&-title') })),
			_react2.default.createElement(
				'div',
				{ className: cx('&-bar-container') },
				_react2.default.createElement('div', {
					className: cx('&-bar', '&-bar-' + kind, _defineProperty({}, '&-bar-' + kind + '-is-pulsed', percentComplete < 100))
				}),
				_react2.default.createElement('div', {
					className: cx('&-bar-overlay'),
					style: { width: 100 - percentComplete + '%' }
				})
			)
		);
	}
});

exports.default = ProgressBar;
